### PR TITLE
logging: fix out of bounds write in log_strdup

### PR DESCRIPTION
**Pull request by jfischer-no**
log_strdup writes out of bounds of a strdup slab.
e.g: CONFIG_LOG_STRDUP_MAX_STRING=46 and
     LOG_STRBUF_STR_SIZE=47 then in the line L:529
sdupl[LOG_STRBUF_STR_SIZE - 1] = '\0';
writes out of bounds because the available buffer space
is only 44 bytes (rounded up to 48 bytes and minus 4 bytes
for the allocated flag).

### CODE REVIEW

File: subsys/logging/log_core.c [hunk 1]
```diff
@@ -525,9 +525,9 @@ char *log_strdup(const char *str)
 	dupl++;
 	sdupl = (char *)dupl;
 
-	strncpy(sdupl, str, CONFIG_LOG_STRDUP_MAX_STRING - 1);
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply subsys/logging/log_core.c]: I find it not immediately understandable with the two macros (LOG_STRBUF_STR_SIZE and CONFIG_LOG_STRDUP_MAX_STRING). Also the real length of the buffer is not used. I just did not see that CONFIG_LOG_STRDUP_MAX_STRING  is also used in the tests.

**Comment by mike-scott** [comment]
@nordic-krch This seems like a good bug-fix.  Would like to see your thoughts.

**Review by jukkar — APPROVED** [review]
Seems to work as expected. I do not see the crashes anymore in #10698. I think we need this urgently as now qemu_x86 + debugging output does not really work.

**Comment by bluespring** [comment]
I also checked #10698 with this patch and no longer see the crash.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**